### PR TITLE
In-person payments: attach level 3 data when capturing payments

### LINKED ISSUE #1962: [Mobile Payments] Add an Endpoint that Returns Level 3 Data for Order
For the Mobile Payments MVP, we need a endpoint (GET?) at something like payments/orders/<ID>/level_three_data that...

- returns the level 3 data that should be passed to a payment intent creation request

Related mobile issues: https://github.com/woocommerce/woocommerce-ios/issues/4297

cc @ctarda @laurendavissmith @bmccotter

### PR DESCRIPTION
**Pull request by kalessil**
Fixes #1962, #1944

#### Changes proposed in this Pull Request

- Changes the way merchant country is provided for level-3 data generation
- Turned out that level-3 data was already supplied when capturing, so we needed only to add the new attribute there

#### Testing instructions

- Ensure merchant account country is US
- Make a regular purchase with authorizing payments setting on
- Capture the payment from the order details page
- Check stripe dashboard logs and ensure level-3 data were part of the capture event and contains the new property (customer_reference)

-------------------

- [ ] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

### CODE REVIEW

File: includes/class-wc-payment-gateway-wcpay.php [hunk 1]
```diff
@@ -1726,12 +1725,12 @@ public function get_level3_data_from_order( string $merchant_country, WC_Order $
 				'discount_amount'     => $discount_amount, // The amount an item was discounted—if there was a sale,for example, as a non-negative integer.
 			];
 		};
-		$items_to_send = array_map( $process_item, $order_items );
 
 		$level3_data = [
 			'merchant_reference' => (string) $order->get_id(), // An alphanumeric string of up to  characters in length. This unique value is assigned by the merchant to identify the order. Also known as an “Order ID”.
+			'customer_reference' => (string) $order->get_id(),
```
Comment (naman03malhotra) [inline includes/class-wc-payment-gateway-wcpay.php:1731]: For my understanding:

From: https://github.com/Automattic/woocommerce-payments/issues/1962#issuecomment-876758134
>customer_reference. It is the only field in the docs we don't include in the plugin level3 handling. I checked with Karl/Stripe and he assures us we can ignore that field.

A bit confused from the comments, do we need to include this field? as Allen suggested we can ignore it.

File: includes/class-wc-payment-gateway-wcpay.php [hunk 2]
```diff
@@ -1683,11 +1683,11 @@ public function cancel_authorization( $order ) {
 	/**
 	 * Create the level 3 data array to send to Stripe when making a purchase.
 	 *
-	 * @param string   $merchant_country The merchant country.
 	 * @param WC_Order $order The order that is being paid for.
 	 * @return array          The level 3 data to send to Stripe.
 	 */
-	public function get_level3_data_from_order( string $merchant_country, WC_Order $order ): array {
+	public function get_level3_data_from_order( WC_Order $order ): array {
+		$merchant_country = $this->account->get_account_country();
```
Comment (naman03malhotra) [inline includes/class-wc-payment-gateway-wcpay.php:1690]: 💯  👍

**Review by naman03malhotra — COMMENTED** [review]
Looks good! Works well using the testing instructions.
Left a few nits and questions for my understanding.

From the issue description:
>For the Mobile Payments MVP, we need an endpoint (GET?) at something like payments/orders//level_three_data

`/order_id/capture_terminal_payment` will be that endpoint? or we'll be creating a new endpoint for this. Do we need to create a separate issue to track this?

---
From: https://github.com/Automattic/woocommerce-payments/issues/1962#issuecomment-879264226
> Karl/Stripe says it is OK to add level3 at the time of capture

We are presently sending (in `develop`) level3 data during payment capture (without the new field), correct?

---

**Comment by kalessil** [comment]
@naman03malhotra : it turned out that we are ok to add the level-3 data in this endpoint, instead of creating the new one. And the data has been injected already in the `capture_charge` method. Hence I kept only the refactoring part I did on the way.

File: includes/class-wc-payment-gateway-wcpay.php (hunk 1 above)
Comment (kalessil) [reply includes/class-wc-payment-gateway-wcpay.php:1731]: We definitely should, as it's a reference for buyers.

File: includes/class-wc-payment-gateway-wcpay.php [hunk 3]
```diff
@@ -1726,12 +1725,12 @@ public function get_level3_data_from_order( string $merchant_country, WC_Order $
 				'discount_amount'     => $discount_amount, // The amount an item was discounted—if there was a sale,for example, as a non-negative integer.
 			];
 		};
-		$items_to_send = array_map( $process_item, $order_items );
```
[the remark this answers is not in the document]
Comment (kalessil) [reply includes/class-wc-payment-gateway-wcpay.php:1729]: In this case, there are no best practices, I'll ever this.

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 4]
```diff
@@ -66,24 +66,21 @@ public function register_routes() {
 	 */
 	public function capture_terminal_payment( $request ) {
 		try {
-			$intent_id = $request['payment_intent_id'];
```
[the remark this answers is not in the document]
Comment (kalessil) [reply includes/admin/class-wc-rest-payments-orders-controller.php:69]: Both ways will work, let's do as you suggested =)

File: includes/class-wc-payment-gateway-wcpay.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (kalessil) [reply includes/class-wc-payment-gateway-wcpay.php]: Fixed in d2747944b21fe0487fce5182284df1454875a9ca

File: includes/admin/class-wc-rest-payments-orders-controller.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (kalessil) [reply includes/admin/class-wc-rest-payments-orders-controller.php:69]: Fixed in d2747944b21fe0487fce5182284df1454875a9ca

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 5]
```diff
@@ -66,24 +66,21 @@ public function register_routes() {
...
-			$order_id  = $request['order_id'];
```
[the remark this answers is not in the document]
Comment (kalessil) [reply includes/admin/class-wc-rest-payments-orders-controller.php:70]: Fixed in d2747944b21fe0487fce5182284df1454875a9ca

**Review by naman03malhotra — APPROVED** [review]
LGTM! :shipit:

**Comment by allendav** [comment]
Thank you @kalessil ! cc @ctarda

**Pull request merged**